### PR TITLE
[ticket/13981] Add events to capture avatar deletion or overwriting

### PR DESCRIPTION
**Pull request by javiexin**
PHPBB3-13981

### CODE REVIEW

File: phpBB/phpbb/avatar/driver/upload.php [hunk 1]
```diff
@@ -80,6 +80,8 @@ public function prepare_form($request, $template, $user, $row, &$error)
 	*/
 	public function process_form($request, $template, $user, $row, &$error)
 	{
+		global $phpbb_dispatcher;
```
Comment (marc1706) [inline phpBB/phpbb/avatar/driver/upload.php]: The phpbb_dispatcher should be passed as a service and not be global.

File: phpBB/phpbb/avatar/driver/upload.php [hunk 2]
```diff
@@ -185,10 +218,33 @@ public function prepare_form_acp($user)
 	*/
 	public function delete($row)
 	{
+		global $phpbb_dispatcher;
```
Comment (marc1706) [inline phpBB/phpbb/avatar/driver/upload.php]: Same as above

File: phpBB/phpbb/avatar/driver/upload.php (hunk 1 above)
Comment (javiexin) [inline phpBB/phpbb/avatar/driver/upload.php]: Fixed.

File: phpBB/phpbb/avatar/driver/upload.php [hunk 3]
```diff
@@ -31,15 +36,17 @@ class upload extends \phpbb\avatar\driver\driver
 	* @param string $php_ext PHP file extension
 	* @param \phpbb_path_helper $path_helper phpBB path helper
 	* @param \phpbb\mimetype\guesser $mimetype_guesser Mimetype guesser
+	* @param \phpbb\event\dispatcher_interface $dispatcher phpBB Event dispatcher object
 	* @param \phpbb\cache\driver\driver_interface $cache Cache driver
 	*/
-	public function __construct(\phpbb\config\config $config, $phpbb_root_path, $php_ext, \phpbb\path_helper $path_helper, \phpbb\mimetype\guesser $mimetype_guesser, \phpbb\cache\driver\driver_interface $cache = null)
+	public function __construct(\phpbb\config\config $config, $phpbb_root_path, $php_ext, \phpbb\path_helper $path_helper, \phpbb\mimetype\guesser $mimetype_guesser, \phpbb\cache\driver\driver_interface $dispatcher, \phpbb\cache\driver\driver_interface $cache = null)
```
Comment (marc1706) [inline phpBB/phpbb/avatar/driver/upload.php]: The dispatcher is not \phpbb\cache\driver\driver_interface

Comment (javiexin) [inline phpBB/phpbb/avatar/driver/upload.php]: Oops... Copy&paste creep.  Fixed.

Comment (javiexin) [inline phpBB/phpbb/avatar/driver/upload.php]: Oops... Copy&paste creep.  Fixed.

**Comment by javiexin** [comment]
Are the Travis and Bamboo errors relevant, or just caused because we have changed the constructor parameters, and therefore the tests do not match the new prototype?

**Comment by marc1706** [comment]
No, you need to update the constructor(s) of the upload avatar in the tests. Travis is telling you which files you need to look in.

**Comment by javiexin** [comment]
I don't know how to do that, I have never used Travis myself.  I will look into it though.

**Comment by javiexin** [comment]
These tests failing seem unrelated to the submitted PR (

/home/travis/build/phpbb/phpbb/phpBB/phpbb/event/md_exporter.php:304
/home/travis/build/phpbb/phpbb/phpBB/phpbb/event/md_exporter.php:145
/home/travis/build/phpbb/phpbb/tests/event/md_exporter_test.php:121

**Comment by Zoddo** [comment]
@javiexin: Probably related to #3789

File: phpBB/phpbb/avatar/driver/upload.php [hunk 4]
```diff
@@ -151,13 +167,35 @@ public function process_form($request, $template, $user, $row, &$error)
...
+		* @var	array	error				Array of errors, if filled in by this event file will not be moved
+		* @since 3.1.6-RC1
+		*/
+		$vars = array(
+			'destination',
+			'prefix',
+			'row',
+			'error',
+		);
+		extract($this->dispatcher->trigger_event('core.avatar_driver_upload_overwrite_before', compact($vars)));
```
Comment (marc1706) [inline phpBB/phpbb/avatar/driver/upload.php]: I'd like it more if you could change the name to core.avatar_driver_upload_move_file_before. It's not always overwriting a file so calling it that is somewhat not correct.

Comment (javiexin) [inline phpBB/phpbb/avatar/driver/upload.php]: Ok, change is coming.

**Pull request merged**